### ISSUE TITLE
Empty "correctResponsesPattern" array

### ISSUE THREAD

**Issue by brianjmiller** [adlnet/xAPI-Spec#698]
Can the value of the `correctResponsesPattern` property of an Interaction Activity Definition contain an empty array (`[]`)? If it can, what is the interpretation of that value?

**Comment by canweriotnow** [comment]
Anything you say is wrong?

**Comment by fugu13** [comment]
Empty should be fine, at least for certain interaction types; SCORM 2004 RTE allows for it (with specified meaning, the correct response is no choice) in at least multiple choice interactions. I'd recommend we just allow for it in general.

**Comment by garemoko** [comment]
Just to clarify what we're talking about here, I think Brian is asking about:
```
"correctResponsesPattern": [ ]
```
This wasn't really possible with SCORM as you either set correct_responses.0.pattern or you didn't. presumably you couldn't set correct_responses to be an empty array.

What SCORM 2004 4th Ed. Table 4.2.9.1a is talking about corresponds to (I think!):
```
"correctResponsesPattern": [
        ""
    ]
```

The other related scenario is that the result object does not have a correctResponsesPattern property, which I take to mean that the correctResponsesPattern is undefined; we're not told the right answer and there may or may not be one. 

Given that, there's two remaining options for what ```"correctResponsesPattern": [ ]``` could mean:
1. There are no right answers - whatever you do (even selecting no options) is wrong. 
2. There are no wrong answers, any response is right. 

The arguments for option 2 are that this is a much more likely scenario and that it follows the spirit of the rules about an empty string as a correctResponsesPattern. 

The argument for option 1 is that theoretically you could represent option 2 by listing all the possible ways of answering the question, but there's no other way to represent option 1. 

Is anybody aware of any adoption either deliberately or accidentally using ```"correctResponsesPattern": [ ]```. 

Given the ambiguity that exists today, in practical terms a tool reading the data could determine the intended meaning of ```"correctResponsesPattern": [ ]``` by looking if success was true or false. Of course, this only works for statements that have a result success property and doesn't help when looking at the activity definition on its own.

**Comment by toaster33** [comment]
Referring to [Testing Requirements](https://github.com/adlnet/xAPI_LRS_Test/blob/master/TestingRequirements.md):
* A "contextActivities" property contains one or more key/value pairs (Format, 4.1..a, 4.1.6.2.b)

My interpretation was is must have a key / value.

**Comment by garemoko** [comment]
That sounds like a reasonable interpretation of the testing document, but the testing document is supposed to be based on the spec and I don't think that's in the spec or should be in the spec.

**Comment by toaster33** [comment]
Fair enough, then referring back to [spec] (https://github.com/adlnet/xAPI-Spec/blob/master/xAPI.md#details-15):
* There are four valid context types. All, any or none of these MAY be used in a given Statement:

Seems to say contextActivities could be provided by a valid context type is not required.

**Comment by garemoko** [comment]
Wait, sorry I misread your first message. This thread is about correctResponsesPattern not contextActivities.

**Comment by brianjmiller** [comment]
> Is anybody aware of any adoption either deliberately or accidentally using `"correctResponsesPattern": [ ]`.

Yes, it came in from a support ticket. Our LRS accepts it (not saying that is correct, just stating a fact) but Wax rejects it (not saying that is incorrect, just stating a fact). I can't say whether it was accidental or deliberate.

**Comment by toaster33** [comment]
[Reference](http://scorm.com/scorm-explained/technical-scorm/run-time/run-time-reference/):
* cmi.interactions.n.correct_responses.n.pattern (format depends on interaction type, RW) One correct response pattern for the interaction.

I think we allow is also (not saying if right either ;).

**Comment by garemoko** [comment]
@toaster33 so the property in the spec we're discussing relates to 'cmi.interactions.n.correct_responses', not  'cmi.interactions.n.correct_responses.n.pattern' 

That's basically what I was trying to say [above](https://github.com/adlnet/xAPI-Spec/issues/698#issuecomment-100176751).

**Comment by toaster33** [comment]
SCORM does have in parenthesis "non-negative integer."

**Comment by canweriotnow** [comment]
@garemoko 
> Is anybody aware of any adoption either deliberately or accidentally using `"correctResponsesPattern": [ ]`

Well, assuming (and why not) that xAPI is in use by Starfleet Academy in the 2280s, the _[Kobayashi Maru](http://en.wikipedia.org/wiki/Kobayashi_Maru)_ scenario comes to mind.

I know the reference is science fiction, but I think the possibility of the set of correct responses being the empty set is an interesting possibility and one worth preserving; some scenarios are negative-sum.

Maybe later a representation of the union of all sets can be devised for "any response is correct."

**Comment by garemoko** [comment]
Hopefully Starfleet will be using server side tracking to prevent Kirk hacking the statement stream...

**Comment by canweriotnow** [comment]
![kirk-xapi](https://cloud.githubusercontent.com/assets/387693/7541374/e9ca4bf6-f582-11e4-8859-955959eb1597.png)

**Comment by canweriotnow** [comment]
@garemoko Starfleet is not yet involved, let's not rely on 22nd century technology to ensure the integrity of this standard, mmmkay?

**Comment by garemoko** [comment]
On call Oct 7th 2015 @fugu13 says - there can be an empty array. The interpretation is that no response is correct because we say that the response is only correct if is matches a pattern in the array. 

Nobody else came down on either side of the argument. 

Conclusion: Add a note to that effect to the relevant section.

**Comment by garemoko** [comment]
@brianjmiller can this issue be closed now that #774 has been merged?
